tests/components/homematicip_cloud/test_hap.py

```diff
@@ -121,3 +150,81 @@
     await hap.async_reset()
 
     assert len(hass.config_entries.async_forward_entry_unload.mock_calls) == 8
+
+
+async def test_hap_create(hass, hmip_config_entry, simple_mock_home):
+    """Mock AsyncHome to execute get_hap."""
+    hass.config.components.add(HMIPC_DOMAIN)
+    # One platform should be enough.
+    const.COMPONENTS = {"binary_sensor"}
+    hap = HomematicipHAP(hass, hmip_config_entry)
+    assert hap
+    with patch.object(
+        hap, "_create_home", return_value=mock_coro(simple_mock_home)
+    ), patch.object(hap, "async_connect", return_value=mock_coro(None)):
+        assert await hap.async_setup() is True
+
+    hass.data[HMIPC_DOMAIN] = {HAPID: hap}
+    await hass.async_block_till_done()
+
+
+async def test_hap_create_exception(hass, hmip_config_entry, simple_mock_home):
+    """Mock AsyncHome to execute get_hap."""
+    hass.config.components.add(HMIPC_DOMAIN)
+    hap = HomematicipHAP(hass, hmip_config_entry)
+    assert hap
+    try:
+        with patch.object(hap, "get_hap", side_effect=HmipConnectionError):
+            await hap.async_setup()
+    except HmipConnectionError as ex:
+        assert isinstance(ex, HmipConnectionError)
+
+    try:
+        simple_mock_home.init.side_effect = HmipConnectionError
+        with patch.object(
+            hap, "_create_home", return_value=mock_coro(simple_mock_home)
+        ):
+            await hap.async_setup()
+    except HomeAssistantError as ex:
+        assert isinstance(ex, ConfigEntryNotReady)
+
+
+async def test_auth_create(hass, simple_mock_auth):
+    """Mock AsyncAuth to execute get_auth."""
+    config = {
+        const.HMIPC_HAPID: HAPID,
+        const.HMIPC_PIN: HAPPIN,
+        const.HMIPC_NAME: "hmip",
+    }
+    hmip_auth = HomematicipAuth(hass, config)
+    assert hmip_auth
+
+    with patch.object(
+        hmip_auth, "_create_auth", return_value=mock_coro(simple_mock_auth)
```

Comment:
Instead of patching our code, we should patch the library code to return the mock we created.

Reply:
`_create_auth` is a method in our code. So if we patch that, it won't be covered in the test. Instead of patching that method, we should patch the library class that `_create_auth` creates an instance of.

Reply:
I tried something like that:
```
    with patch(
        "homematicip.aio.connection.AsyncConnection", autospec=True
    ), patch(
        "homematicip.aio.home.AsyncHome", autospec=True, side_effect=mock_coro(simple_mock_home)
    ), patch(
        "homematicip.aio.auth.AsyncAuthConnection", autospec=True
    ), patch(
        "homematicip.aio.auth.AsyncAuth", autospec=True, side_effect=mock_coro(simple_mock_auth)
    ):
        yield
```
, but with no success.

Reply:
`AsyncAuth` is a class so should just return an instance of that class, right? So we want the patch of that class to return this mock instance: `Mock(spec=AsyncAuth, pin=HAPPIN, create=True)`.

Reply:
Paragraph one is covered by the fixture simple_mock_auth
```
@pytest.fixture(name="simple_mock_auth")
def simple_mock_auth_fixture():
    """Return a simple AsyncAuth Mock."""
    return Mock(spec=AsyncAuth, pin=HAPPIN, create=True)
```
test_hap.py
```
    with patch(
        "homematicip.aio.auth.AsyncAuth", return_value=simple_mock_auth
    ):
        assert await hmip_auth.async_setup() is True
```
right?

Reply:
I think my head is blocked. Please add code.

Reply:
I think that looks good. Although `simple_mock_auth_fixture` is currently a coroutine, which it doesn't need to be I think.

Reply:
I removed the coroutine from `simple_mock_auth_fixture`

With regards to line 199 in `test_hap.py`:

This patch works: 
`with patch.object(hmip_auth, "_create_auth", return_value=simple_mock_auth):`

This patch still does not work:
`with patch("homematicip.aio.auth.AsyncAuth", return_value=simple_mock_auth):`

`_create_auth` wont't return a mock with the secound version.

Reply:
We need to patch in the correct place, use `"homeassistant.components.homematicip_cloud.hap.AsyncAuth" `.

https://docs.python.org/3/library/unittest.mock.html#where-to-patch

Reply:
Martin, you made my day!
That was the tip i was searching for a long time now.
Thanks for your patience.